### PR TITLE
Intan digital channel extraction for header attached and one file per signal format

### PR DESCRIPTION
**Pull request by h-mayorquin**
@zm711 this is a follow-up of  https://github.com/NeuralEnsemble/python-neo/pull/1476 

This makes the extraction of the digital channels work like all the other data in neo (and therefore spikeinterface) so user does not need to do any post-processing. The post-processing was derived from the data sheet, the python [conversion script that intan provides for python](https://intantech.com/downloads.html?tabSelect=Software&yPos=68.57142639160156) and [the matlab version](https://github.com/dattalab/spike_finder/blob/master/read_Intan_RHD2000_file.m) of the same operation .

Let me know what you think.

@luiztauffer

I tested this on Nelson data like this (using this branch and from spikeinterface):

```python
from pathlib import Path

intan_data_folder = Path.home() / "Downloads" 
assert intan_data_folder.is_dir()

file_path = intan_data_folder / "pm231213a_240205_240205_152654.rhd"
assert file_path.is_file()


from spikeinterface.extractors import IntanRecordingExtractor

recording = IntanRecordingExtractor(file_path=file_path, stream_name="USB board digital input channel", all_annotations=True)
recording
```
<div style='border:1px solid #ddd; padding:10px;'><strong>IntanRecordingExtractor: 3 channels - 30.0kHz - 1 segments - 1,800,064 samples - 60.00s (1.00 minutes) - uint16 dtype - 10.30 MiB</strong></div><details style='margin-left: 10px;'>  <summary><strong>Channel IDs</strong></summary><ul>['DIGITAL-IN-12' 'DIGITAL-IN-13' 'DIGITAL-IN-14'] </details><details style='margin-left: 10px;'>  <summary><strong>Annotations</strong></summary><ul><li> <strong> is_filtered </strong>: False</li><li> <strong> name </strong>: USB board digital input channel</li><li> <strong> stream_id </strong>: 4</li><li> <strong> file_origin </strong>: /home/heberto/Downloads/pm231213a_240205_240205_152654.rhd</li><li> <strong> intan_version </strong>: 3.3</li><li> <strong> actual_upper_bandwidth </strong>: 7603.76513671875</li><li> <strong> desired_lower_bandwidth </strong>: 0.10000000149011612</li><li> <strong> note1 </strong>: </li><li> <strong> actual_impedance_test_frequency </strong>: 1000.0</li><li> <strong> notch_filter_mode </strong>: 0</li><li> <strong> num_temp_sensor_channels </strong>: 0</li><li> <strong> desired_upper_bandwidth </strong>: 7500.0</li><li> <strong> desired_impedance_test_frequency </strong>: 1000.0</li><li> <strong> note2 </strong>: </li><li> <strong> nb_signal_group </strong>: 7</li><li> <strong> actual_dsp_cutoff_frequency </strong>: 1.165827989578247</li><li> <strong> eval_board_mode </strong>: 0</li><li> <strong> note3 </strong>: </li><li> <strong> desired_dsp_cutoff_frequency </strong>: 1.0</li><li> <strong> actual_lower_bandwidth </strong>: 0.09452909976243973</li><li> <strong> notch_filter </strong>: False</li><li> <strong> dsp_enabled </strong>: 1</li></ul> </details><details style='margin-left: 10px;'><summary><strong>Channel Properties</strong></summary><ul><details><summary> <strong> gain_to_uV </strong> </summary>[1. 1. 1.]</details><details><summary> <strong> offset_to_uV </strong> </summary>[0. 0. 0.]</details><details><summary> <strong> channel_names </strong> </summary>['DIGITAL-IN-12' 'DIGITAL-IN-13' 'DIGITAL-IN-14']</details><details><summary> <strong> channel_ids </strong> </summary>['DIGITAL-IN-12' 'DIGITAL-IN-13' 'DIGITAL-IN-14']</details><details><summary> <strong> native_order </strong> </summary>[12 13 14]</details><details><summary> <strong> chip_channel_num </strong> </summary>[0 0 0]</details><details><summary> <strong> electrode_impedance_phase </strong> </summary>[0. 0. 0.]</details><details><summary> <strong> spike_scope_digital_trigger_channel </strong> </summary>[0 0 0]</details><details><summary> <strong> electrode_impedance_magnitude </strong> </summary>[0. 0. 0.]</details><details><summary> <strong> spike_scope_digital_edge_polarity </strong> </summary>[0 0 0]</details><details><summary> <strong> spike_scope_trigger_mode </strong> </summary>[1 1 1]</details><details><summary> <strong> board_stream_num </strong> </summary>[0 0 0]</details><details><summary> <strong> custom_order </strong> </summary>[12 13 14]</details><details><summary> <strong> spike_scope_voltage_thresh </strong> </summary>[0 0 0]</details></ul></details>

```
full_traces = recording.get_traces()
partial_traces = recording.get_traces(start_frame=0, end_frame=3_000, channel_ids=["DIGITAL-IN-12", "DIGITAL-IN-14" ])

np.allclose(full_traces[0:3_000, [0, 2]], partial_traces)
```

Can you check if the output makes sense in the light of the experiment you are converting? The signal looks correct but nothing like ecological validation.

### CODE REVIEW

**Comment by zm711** [comment]
Cool I'll look it over soon. What's the output? Just 1 and 0s? right?

**Comment by h-mayorquin** [comment]
> Cool I'll look it over soon. What's the output? Just 1 and 0s? right?

image


Yes, I was torn between that and bools but at the end I decided to use np.int8 as I don't know if traces are returned as bool for any other case so I rather play with that expectation.

**Comment by zm711** [comment]
Cool that makes sense. I know Intan let's you choose, but I think 1 and 0s are probably the better expectation.

File: neo/rawio/intanrawio.py
Comment (zm711) [inline neo/rawio/intanrawio.py:208]:
```suggestion
        self.native_channel_order = {}
```

I know others like `dict()`, but `{}` is faster. [review here](https://madebyme.today/blog/python-dict-vs-curly-brackets/)

I like {} better, but since the performance is only about 2x faster if you really prefer dict() we can keep it.

File: neo/rawio/intanrawio.py
Comment (zm711) [inline neo/rawio/intanrawio.py:347]: Remind me the `[:]` I always forget this :)

File: neo/rawio/intanrawio.py
Comment (zm711) [inline neo/rawio/intanrawio.py:350]: What is a field name? 

I wonder if we will remember this in the future. It is not:
1) Neo terminology (which we are bound to here--block, stream, segment)
2)  Not clear to me as a word in this case

File: neo/rawio/intanrawio.py
Comment (zm711) [inline neo/rawio/intanrawio.py:651]: Nice renaming!

File: neo/rawio/intanrawio.py
Comment (zm711) [inline neo/rawio/intanrawio.py:1124]: Technically we could just check `stream_name in name_to_field_name_digital` and get rid of line 1123, but you are going for clarity over reduced lines I think.

I would still prefer to think if we have a better naming than `name_to_field_name_digital`, but we are using so many terms maybe this is the best.

**Review by zm711 — COMMENTED** [review]
First set of comments on my side.

File: neo/rawio/intanrawio.py
Comment (h-mayorquin) [reply neo/rawio/intanrawio.py:347]: You mean the why? One is a reference the other gets the data which I need here to have the string version.

File: neo/rawio/intanrawio.py
Comment (h-mayorquin) [reply neo/rawio/intanrawio.py:350]: This is standard terminology for struct and structured arrrays in numpy:

https://numpy.org/doc/stable/glossary.html#term-field

File: neo/rawio/intanrawio.py
Comment (h-mayorquin) [reply neo/rawio/intanrawio.py:651]: Maybe we can go further as there are other references to type instead of id in the code but I wanted to keep the diff small for the first review.

File: neo/rawio/intanrawio.py
Comment (zm711) [reply neo/rawio/intanrawio.py:350]: Okay perfect! You're right. I never use structured arrays, but Neo and SI use them all the time, so I should get more used to the naming. Maybe it should be 

```
stream_name_to_field_name_digital
```

To be as explicit as possible?

File: neo/rawio/intanrawio.py
Comment (h-mayorquin) [reply neo/rawio/intanrawio.py:1124]: The name is bad, I agree. Plus, this is bad practice because there is an implicit coupling between the values here and the lines here:

https://github.com/h-mayorquin/python-neo/blob/2af13478edbe92e55a8d308ef7fc8a29220ada40/neo/rawio/intanrawio.py#L799-L801

In other words, a relationship that is implicit. That should not be there.

File: neo/rawio/intanrawio.py
Comment (h-mayorquin) [reply neo/rawio/intanrawio.py:208]: I will change this on the next commit.

File: neo/rawio/intanrawio.py
Comment (h-mayorquin) [reply neo/rawio/intanrawio.py:350]: Good idea. Let's start with this.

File: neo/rawio/intanrawio.py
Comment (h-mayorquin) [reply neo/rawio/intanrawio.py]: OK, I decided that for this cases is better to use the stream_name as the field name (as all the data is packed in one field). This avoids us having to define an artificial channel name (as we were doing) and sidesteps the need to define such mapping between stream and the artificial channel name. It also works nicely with the case of one-file-per-signal as the signal is the stream.

File: neo/rawio/intanrawio.py
Comment (h-mayorquin) [reply neo/rawio/intanrawio.py]: OK, I went further and changed signal_type for stream_id in most places. Check it out.

File: neo/rawio/intanrawio.py
Comment (h-mayorquin) [reply neo/rawio/intanrawio.py]: I decided what I think is a simpler approach. See below.

File: neo/rawio/intanrawio.py
Comment (h-mayorquin) [reply neo/rawio/intanrawio.py:347]: Done.

File: neo/rawio/intanrawio.py
Comment (h-mayorquin) [reply neo/rawio/intanrawio.py]: Done.

**Comment by zm711** [comment]
I'll re-check this out tomorrow. I selfishly need to refactor some of my own code that relies on the old way to return the digital streams. So I will load this on my workstation and test for our uses locally too!

**Comment by h-mayorquin** [comment]
OK, thanks, that will be really useful.

**Comment by samuelgarcia** [comment]
I do not have much for reading this in detail.
My first impression is I do not like too much.
You need it now so lets go and merge it now  but...
...but...
you need to swear that when we implement this famous get_digitalsignal_chunk that is never coming you will help me to revert this back.

**Comment by zm711** [comment]
Yep of course. I have a bunch of test data locally and I can create a simulated file for gin to test all of this for now so we will make sure it is all good before merging :)

**Comment by h-mayorquin** [comment]
Hey Sam, I think it is important for users to be able to extract this information directly without having to rely on formatk knowledge to use it. 

If you have some time please comment on what you don't like about it? Maybe there is a way of addressing that.

**Comment by zm711** [comment]
The problem isn't with how we are doing it. The goal is to have analog data and then digital data. But the way to access the digital data isn't built into Neo yet. So Sam is saying that once we have the digital machinery at the BaseRawIO level we should move all the digital-in and out logic there instead of inside the analog machinery where it technically doesn't belong.

**Comment by samuelgarcia** [comment]
Sorry. Thanks a lot Zach for this extacts translation. This was exactly what I meant.

**Comment by luiztauffer** [comment]
@h-mayorquin @zm711 I just tested it, and it seems to be working fine! Thanks for working on this and sorry for taking me a couple of days to try it out
can we merge it and make a release? this will be needed for TTL-based synchronization for every lab using Intan recordings

**Comment by zm711** [comment]
Thanks for testing @luiztauffer. 

I was planning on testing as well last week, but I will do my best to test this tomorrow. I'll bring up release at our next meeting for this.

**Comment by zm711** [comment]
So I tested this and it seems to work for me. The only thing is that for groups that have already hacked the memmap to pull out digital data using the field name 'DIGITAL-IN' will now have to use `USB board digital input channel`. But seems fine. Let me read through the code and then this is good to merge unless you have any ideas to help prevent breaking the code @h-mayorquin ?

File: neo/rawio/intanrawio.py
Comment (CodyCBakerPhD) [reply neo/rawio/intanrawio.py:775]: Any SI unit would be preferable, including "n.a." or "a.u."

Comment (zm711) [reply neo/rawio/intanrawio.py:775]: I'm absolutely neutral on this. As long as we don't give  real units like volts/millivolt etc. So if that helps with your guys' schemata then I'm fine with @h-mayorquin changing this to "a.u." if he wants.

File: neo/rawio/intanrawio.py
Comment (h-mayorquin) [reply neo/rawio/intanrawio.py:657]: By all means please do, that was an omission of mine caused by ignorance.

File: neo/rawio/intanrawio.py
Comment (h-mayorquin) [reply neo/rawio/intanrawio.py:775]: We don't really use units as far as I remember but I think @CodyCBakerPhD idea is a good one.  I will change it.

**Comment by h-mayorquin** [comment]
> So I tested this and it seems to work for me. The only thing is that for groups that have already hacked the memmap to pull out digital data using the field name 'DIGITAL-IN' will now have to use `USB board digital input channel`. But seems fine. Let me read through the code and then this is good to merge unless you have any ideas to help prevent breaking the code @h-mayorquin ?

Can you show me a quick snippet of how they are accesing it? It is from spikeinterface or neo? I never use the neo layer to access data so I need to think on how the hacking looks like. Maybe there is a way.

File: neo/rawio/intanrawio.py
Comment (h-mayorquin) [reply neo/rawio/intanrawio.py:125]: We sort of need both I guess, one tests the file exists and the other that is of the correct path. Otherwise a non-existen file that ends in rhs might cause a confusion error down the line, right?

**Comment by zm711** [comment]
Sure. Before we exposed this to the user you could still get the data by doing
```python
raw_digital_data = reader._raw_data['DIGITAL-IN']

```
Technically this is all private so we can change this and it is on the user to adapt, but if people were hacking to get this data they would need to change for this. I'm not saying we have to adapt for this because it was private previously but it was easy to make code to hack this.

File: neo/rawio/intanrawio.py
Comment (zm711) [reply neo/rawio/intanrawio.py:125]: Okay. I guess we could also do your boolean and then add a quick exists(), but I think it is also fine to keep them as two separate checks as well.

File: neo/rawio/intanrawio.py
Comment (h-mayorquin) [reply neo/rawio/intanrawio.py:438]: image

When you read the numpy structured arrays you read blocks the blocks as columns.  So you need to reshape this as if you were concatenating all the blocks.

Comment (zm711) [reply neo/rawio/intanrawio.py:438]: I understand the flattening, but why aren't we using the 

`self._raw_data[field_name].flatten()[sl0:sl1]` for the header attached?

**Comment by h-mayorquin** [comment]
Mmm I see, no I rather punish wrong doers for accessing our private thing : P

More seriously, I think we could do something if they were accessing this through the API so we throw a proper deprecation warning. But because they are just accessing the memmap directly there is no way to throw a reasonable deprecation warning I think.

Another option would be to keep two references to the same memmap but then again because there is no deprecation warning that we can throw there is no way to change it slowly. The pain is now or at some other point and we might as well do it now. 

If someone is relying on that I guess they can pin neo to the last version?

File: neo/rawio/intanrawio.py
Comment (h-mayorquin) [reply neo/rawio/intanrawio.py:437]: I think I was lazy to implement this because I thought the memory footprint is small? Would you like to give it a go?

We would need to pass this:

```
            digital_raw_data = self._raw_data[field_name][block_start:block_stop].flatten()[sl0:sl1]
```

And then demuliplex only that piece.

**Comment by h-mayorquin** [comment]
I just realize that I don't have the rights to merge master on the web interface.

File: neo/rawio/intanrawio.py
Comment (zm711) [reply neo/rawio/intanrawio.py:437]: I see what you mean. So you always load the full and then take what you want. I don't think it is necessary. This vector should always be relatively small compared to the amplifier data, so I don't think we need to worry about memory footprint. We can do that later if needed...

**Comment by zm711** [comment]
Once tests pass I'll add the comment and then I'll merge after they pass the final time :)

File: neo/rawio/intanrawio.py
Comment (h-mayorquin) [reply neo/rawio/intanrawio.py]: Mmm I think the two checks point the user to two different problems. I think they should remain separate. You mght have pass the wrong file or your might have passed other file from intan that is not rhs or rhd.  I think we should keep it more or less as it is now.

**Comment by zm711** [comment]
Missing a Path :)

**Comment by h-mayorquin** [comment]
Let me fix it.

**Comment by zm711** [comment]
I'll wait for you to have a chance to read the comments I added and then if you're okay with things, then this should be good to go.

**Review by zm711 — APPROVED** [review]

**Comment by h-mayorquin** [comment]
Comments look good to me.

**Comment by zm711** [comment]
Do you know why tests decided not to run?

**Comment by h-mayorquin** [comment]
It is Friday? No idea

**Comment by zm711** [comment]
Found the setting and turned it on so I'll update this branch but from now Neo should give people the option to click a button to update!

**Pull request merged**
